---
title: "Single-cell Analysis of Host-Microbiome Interactions"
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>",
    fig.path = "man/figures/README-",
    out.width = "100%"
)
```

<!-- badges: start -->
[![CRAN status](https://www.r-pkg.org/badges/version/rsahmi)](https://CRAN.R-project.org/package=rsahmi)
[![Project Status: Active - The project has reached a stable, usable state and is being actively developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![](https://cranlogs.r-pkg.org/badges/rsahmi)](https://cran.r-project.org/package=rsahmi)
[![DOI](https://zenodo.org/badge/605180683.svg)](https://doi.org/10.5281/zenodo.15076675)
<!-- badges: end -->

A computational resource designed to accurately detect microbial nucleic acids
while filtering out contaminants and false-positive taxonomic assignments from
standard transcriptomic sequencing of mammalian tissues (See
[SAHMI](https://github.com/sjdlabgroup/SAHMI)). This implementation leverages
the `polars` package for fast and systematic microbial signal recovery and
denoising from host tissue genomic sequencing.

## Installation
You can install `rsahmi` from `CRAN` using:
```{r, eval=FALSE}
# install.packages("pak")
pak::pak("rsahmi")
```

Alternatively, install the development version from
[r-universe](https://yunuuuu.r-universe.dev/rsahmi) with:

```{r, eval=FALSE}
pak::repo_add("https://yunuuuu.r-universe.dev")
pak::pak("rsahmi")
```

or from [GitHub](https://github.com/Yunuuuu/rsahmi) with:
```{r, eval=FALSE}
pak::pak("Yunuuuu/rsahmi")
```

You must also install [seqkit](https://bioinf.shenwei.me/seqkit/) and
[kraken2](https://github.com/DerrickWood/kraken2/wiki/Manual).

Additionally, `rsahmi` relies on [polars](https://rpolars.github.io/index.html).
If you call any functions that require `polars`, you will be prompted to install
the package.

## Cite
If you use this package, please cite:

 - Song, Yuxuan PhD; Peng, Yun PhD; Qin, Caipeng PhD; Jiang, Shan PhD; Lin,
 Jiaxing PhD; Lai, Shicong MD; Wu, Jilin PhD; Ding, Mengting PhD; Du, Yiqing
 PhD*; Yu, Luping MD*; Xu, Tao MD*. Antibiotic use attenuates response to immune
 checkpoint blockade in urothelial carcinoma via inhibiting CD74-MIF/COPA:
 revealing cross-talk between anti-bacterial immunity and ant-itumor immunity.
 International Journal of Surgery 111(1):p 972-987, January 2025. | DOI:
 10.1097/JS9.0000000000001901.
 - Ghaddar B, Blaser MJ, De S. Denoising sparse microbial signals from
single-cell sequencing of mammalian host tissues. Nat Comput Sci. 2023
Sep;3(9):741-747. doi: 10.1038/s43588-023-00507-1. Epub 2023 Sep 18. PMID:
37946872; PMCID: PMC10634611.
